### PR TITLE
operator: extract controller-runtime integration into its own cell

### PR DESCRIPTION
**Pull request by mhofstetter**
Currently the Gateway API controller (and its reconcilers) is the only controller that is making use of the controller-runtime library.

To prepare its usage for other use-cases (e.g. Ingress #28911), this commit extracts the controller-runtime integration into its own cell by providing the relevant components (Manager & Scheme) to dependent cells.

The new controller-runtime cell is responsible to start the manager itself.

Dependent cells register their reconcilers with the manager and types with the scheme respectively.

### CODE REVIEW

File: operator/pkg/controller-runtime/cell.go
Comment (bimmlerd) [inline operator/pkg/controller-runtime/cell.go]: (nit) with this many parameters, I think a `cell.In` struct reads more nicely

File: operator/pkg/gateway-api/cell.go
Comment (bimmlerd) [inline operator/pkg/gateway-api/cell.go]: (nit) maybe `failed to add type %t to scheme: %w` if there's some way to get a human readable name of the type?

(also applies to the registration in `cell.go`)

File: operator/pkg/gateway-api/cell.go
Comment (bimmlerd) [inline operator/pkg/gateway-api/cell.go:147]: (nit) similarly, here, can the reconciler have a human readable name so that the error says which reconciler failed to be setup?

File: operator/pkg/controller-runtime/cell.go
Comment (bimmlerd) [inline operator/pkg/controller-runtime/cell.go:4]: is the mismatch between directory and pkg name intentional?

File: operator/pkg/gateway-api/cell.go
Comment (bimmlerd) [inline operator/pkg/gateway-api/cell.go]: (nit) probably also time for a `cell.In`.

File: CODEOWNERS
Comment (bimmlerd) [inline CODEOWNERS:375]: I think this commit can be squashed with the commit introducing the pkg

**Review by bimmlerd — APPROVED** [review]
nits, but LGTM if you squash the CODEOWNERs commit into the commit introducing the pkg.

File: operator/pkg/controller-runtime/cell.go
Comment (mhofstetter) [reply operator/pkg/controller-runtime/cell.go]: thanks, done

File: operator/pkg/gateway-api/cell.go
Comment (mhofstetter) [reply operator/pkg/gateway-api/cell.go]: thanks, done

File: CODEOWNERS
Comment (mhofstetter) [reply CODEOWNERS:375]: agree -> squashed

File: operator/pkg/controller-runtime/cell.go
Comment (mhofstetter) [reply operator/pkg/controller-runtime/cell.go:4]: yes it's somehow intentional. It's more or less the same way in the library itself -> https://github.com/kubernetes-sigs/controller-runtime/blob/main/alias.go#L17

File: operator/pkg/gateway-api/cell.go
Comment (mhofstetter) [reply operator/pkg/gateway-api/cell.go]: Would be nice, but since its only a function the output of `%T` would be that much of help `func(*runtime.Scheme) error`

The underlying error will contain enough information, and these type of errors are rare anyway.

File: operator/pkg/gateway-api/cell.go
Comment (mhofstetter) [reply operator/pkg/gateway-api/cell.go:147]: I thought about the same. The reconciler has a name, but unfortunately we're re-using the names between different reconcilers - so they aren't of that much help either.

Will follow up on this in a separate PR.

File: operator/pkg/gateway-api/cell.go
Comment (pippolo84) [reply operator/pkg/gateway-api/cell.go]: What about:

```go
for gv, f := range map[fmt.Stringer]func(s *runtime.Scheme) error{
	gatewayv1.GroupVersion:       gatewayv1.AddToScheme,
	gatewayv1beta1.GroupVersion:  gatewayv1beta1.AddToScheme,
	gatewayv1alpha2.GroupVersion: gatewayv1alpha2.AddToScheme,
} {
	if err := f(scheme); err != nil {
		return fmt.Errorf("failed to add types from %s to scheme: %w", gv, err)
	}
}
```

Anyway, not a blocker, I'll leave this to you.

**Review by pippolo84 — APPROVED** [review]
LGTM, thanks :100:

File: operator/pkg/controller-runtime/cell.go
Comment (meyskens) [inline operator/pkg/controller-runtime/cell.go:30]: nit: could we make a bit more useful description here? Who's controller runtime? Where to use? Here on in the comment above

**Review by meyskens — APPROVED** [review]
Small nits overall LGTM

File: operator/pkg/gateway-api/cell.go
Comment (mhofstetter) [reply operator/pkg/gateway-api/cell.go]: Thanks for the idea @pippolo84 . This works if the package provides the group version as global variable. This is not the case for `clientgoscheme` (aggregation of multiple groupversions) that gets added to the scheme in the `controller-runtime` cell. IMO this is why we should rely on the underlying error to provide enough information for analysing any potential issue.

Update: it might work by using the `clientgoscheme.Scheme` as default scheme. This way we don't have to explicitly register this scheme.

File: operator/pkg/controller-runtime/cell.go
Comment (mhofstetter) [reply operator/pkg/controller-runtime/cell.go:30]: Good idea! I changed the comment a little bit - more or less adding a link to their github project.

I don't want to add anything about _where to use it_. Currently there's no guideline that we want to use the controller-runtime for all controllers in the Cilium Operator.

File: operator/pkg/gateway-api/cell.go
Comment (mhofstetter) [reply operator/pkg/gateway-api/cell.go]: Changed

**Pull request merged**
